tests/test_video.py

```diff
@@ -0,0 +1,113 @@
+
+
+def make_sample_video(path, frames=4, frame_size=(10, 20), fps=25.0):
+    """
+    frame_size is (H, W), only even sides
+    """
+
+    writer = cv2.VideoWriter(path, frameSize=tuple(frame_size[::-1]),
+        fps=float(fps), fourcc=cv2.VideoWriter_fourcc(*'MJPG'))
+
+    for _ in range(frames):
+        # Apparently, only uint8 values are supported, but not floats
+        writer.write(np.ones((*frame_size, 3), dtype=np.uint8) * 255)
+
+    writer.release()
+
+@pytest.fixture(scope='module')
+def fxt_sample_video():
+    with TestDir() as test_dir:
+        video_path = osp.join(test_dir, 'video.avi')
+        make_sample_video(video_path, frame_size=(4, 6))
+
+        yield video_path
+
+class VideoTest:
+    @mark_requirement(Requirements.DATUM_GENERAL_REQ)
+    @scoped
+    def test_can_read_video(self, fxt_sample_video):
+        video = Video(fxt_sample_video)
+        on_exit_do(video.close)
+
+        assert 4 == video.frame_count
+        assert (4, 6) == video.frame_size
+
+    @mark_requirement(Requirements.DATUM_GENERAL_REQ)
+    @scoped
+    def test_can_read_frames_sequentially(self, fxt_sample_video):
+        video = Video(fxt_sample_video)
+        on_exit_do(video.close)
+
+        for idx, frame in enumerate(video):
+            assert frame.size == video.frame_size
+            assert frame.index == idx
+            assert id(frame.video) == id(video)
+            assert frame == np.ones((*video.frame_size, 3)) * 255
+
+    @mark_requirement(Requirements.DATUM_GENERAL_REQ)
+    @scoped
+    def test_can_read_frames_randomly(self, fxt_sample_video):
+        video = Video(fxt_sample_video)
+        on_exit_do(video.close)
+
+        for idx in {1, 3, 2, 0, 3}:
+            frame = video[idx]
+            assert frame.index == idx
+            assert frame == np.ones((*video.frame_size, 3)) * 255
+
+    @mark_requirement(Requirements.DATUM_GENERAL_REQ)
+    @scoped
+    def test_can_skip_frames_between(self, fxt_sample_video):
+        video = Video(fxt_sample_video, step=2)
+        on_exit_do(video.close)
+
+        assert 2 == video.frame_count
+
+    @mark_requirement(Requirements.DATUM_GENERAL_REQ)
+    @scoped
+    def test_can_skip_from_start(self, fxt_sample_video):
+        video = Video(fxt_sample_video, start_frame=1)
+        on_exit_do(video.close)
+
+        assert 3 == video.frame_count
+        assert 1 == next(iter(video)).index
+
+    @mark_requirement(Requirements.DATUM_GENERAL_REQ)
+    @scoped
+    def test_can_skip_from_end(self, fxt_sample_video):
+        video = Video(fxt_sample_video, end_frame=2)
+        on_exit_do(video.close)
+
+        last_frame = None
+        for last_frame in video:
+            pass
+
+        assert 2 == video.frame_count
+        assert 1 == last_frame.index
+
+    @mark_requirement(Requirements.DATUM_GENERAL_REQ)
+    @scoped
+    def test_can_init_frame_count_lazily(self, fxt_sample_video):
+        with mock.patch.object(Video, '_get_frame_count',
```

Comment:
IMO, it would be better to create an actual test video file whose frame count can't be determined in advance and use that:

1. It would serve as a proof that such a situation can actually occur in reality.
2. It would avoid the fragility that comes with mocking internal methods.

Reply:
I agree, but it's not simple to create such video with opencv.

Reply:
I don't mean creating it with OpenCV; I mean creating it offline with whatever tool you can use and adding it to the repository.

Reply:
It looks like it [can happen](https://github.com/opencv/opencv/blob/4.x/modules/videoio/src/cap_ffmpeg_impl.hpp#L1633-L1686), but I failed to create a reproducing example. However, it is simple to obtain an incorrect value here, e.g. in case of variable FPS. Videos look to always have `duration` info, which is used to calculate frame count, but it can be rounded up, giving extra frames in computations. Thus, I'm not sure we should provide or use this info until we iterate over frames or have a reliable source of this info.

Reply:
> It looks like it can happen, but I failed to create a reproducing example. However, it is simple to obtain an incorrect value here, e.g. in case of variable FPS. 

Okay, but can you produce an example of that? It could be a useful testcase too.

Reply:
I have a video, for which we have 2 extra frame in computations, however, I was not able to craft such example. Also, we can't just store a random video from internet in the repo for legal and storage reasons. We definitely know it happens sometimes and there are multiple evidence for it across internet. I think, we can just leave a comment in the code.